### ISSUE TITLE
Ensure/check single version of a dependency package throughout workspace

### ISSUE THREAD

**Issue by wclr** [pnpm/pnpm#2713]
I think it is quite a common issue, if you have multiple packages/app in you monorepo that may depend on the same packge dependency (for as common example `mongodb`), it is reasonable to have only single version thought the project (though there maybe need for exeptions sometimes). But eventually you may add this dep to new package, and it will install newer version, but you may not notice this and forget to update the rest. 

What is the best way to approach this issue? I thought about using hook, to check lockfile after resolution. Another approach would be declare such dep in the root of the project, so everyone will use it, but even if all your packages/apps (which is rare case) depend on this, this still doesn't seem a nice, clean and lean solution.

**Comment by zkochan** [comment]
What if we will add something like a linter for `package.json` files. For instance, we can have something like this in a `pnpmfile.js` file:

```js
module.exports = {
  lintPackage (pkg) {
    if (pkg.dependencies?.mongodb) {
      pkg.dependencies.mongodb = '^3.0.0'
    }
    return pkg
  }
}
```

pnpm can use this for two things.

1. it can run this function on each project's manifest and verify that the object return by the function equals the one that was passed in. If they do not match, there is a linting error
2. it can fix packages by writing the returned manifests to the filesystem. So we may have a command like `pnpm fix`.

Yarn v2 has something similar. It is called constraints and they use Prolog for describing these rules.

cc @pnpm/collaborators

**Comment by aparajita** [comment]
Can't `readPackage` do this already?

**Comment by zkochan** [comment]
The changes of readPackage are not saved to the file system

**Comment by aparajita** [comment]
Another possible approach:

- Add a setting to .npmrc, something like `pinWorkspacePackages = <spec>`, where `spec` is a comma-delimited list of package names which can use `*` as a wildcard. `*` alone would match all packages.

- When `pnpm install/update` is executed, `pinWorkspacePackages` is consulted. If any existing packages elsewhere in the workspace match `<spec>` and have the same name as the one being installed/updated and have a different version, they are updated to the same version.

**Comment by whitecolor** [comment]
> Also, I am not sure it should be part of pnpm. It can be probably a separate project that supports pnpm, npm, yarn.

I think such logic should be out of scrope of pnpm itself, but pnpm should allow to implement such linting things using kind of `plugins` (which I belive will be mostly implemented via hooks).

> What if we will add something like a linter for package.json files. 

```
pkg.dependencies.mongodb = '^3.0.0'
```
But does it prevent one package from having `3.1.2` and another `3.3.0` installed later and which are fixed in lock file and not updated non explicitly? I'm not sure how it works, but I belive when lockfile is used installed dep exact version if fixed just after install and will not change until explicit update operation on this package?

**Comment by aparajita** [comment]
> I think such logic should be out of scrope of pnpm itself

pnpm is a... package manager. Which excels at working with monorepos. How is managing package versions across the monorepo out of the scope of pnpm?

**Comment by whitecolor** [comment]
> > I think such logic should be out of scrope of pnpm itself
> 
> pnpm is a... package manager. Which excels at working with monorepos. How is managing package versions across the monorepo out of the scope of pnpm?

Basic logic of `pnpm` should be as simple as possible, should concern such things which is not achievalbe in userland, hooks API are great things which potentially can allow many things out of scope of basic packages resolution/linking logic which is pnpm responsible for.

**Comment by aparajita** [comment]
> Basic logic of pnpm should be as simple as possible

It's a little too late for that. 😂

**Comment by whitecolor** [comment]
> > Basic logic of pnpm should be as simple as possible
> 
> It's a little too late for that. 😂

Well what things pnpm does which could be actually achivable by user using avaialbe API, but not innner pnpm logic?

**Comment by ExE-Boss** [comment]
Didn’t `resolution‑strategy` use to do this?
```properties
resolution-strategy = fewer-dependencies
```

**Comment by zkochan** [comment]
We need two things:

1. packages with the same spec should always be resolved to the same version. So if one project has foo@^3.0.0 and another one has foo@^3.0.0, in both projects pnpm should resolve foo to the same version, to 3.1.0, for instance. This is how Yarn works but Yarn does it on all levels (in subdependencies as well), we'll probably guarantee this only for project dependencies for now. This probably needs a new issue.
2. some linting solution should be possible for pnpm monorepos. It can be a separate tool to integrated to pnpm. Maybe we will just provide a high level API.

**Comment by zkochan** [comment]
> Didn’t resolution‑strategy use to do this?

no. The change introduces by fewer-dependencies was that when dependencies were resolved, versions of already resolved packages from the parent packages were preferred.

**Comment by bdchauvette** [comment]
I've found [`syncpack`](https://www.npmjs.com/package/syncpack) to be pretty handy for ensuring version consistency and standard formatting for package.json files.

It would be great if it had first-party support for `pnpm` workspaces, but globbing works ok.

**Comment by aparajita** [comment]
> I've found syncpack to be pretty handy

Thanks for the tip!

**Comment by aparajita** [comment]
I'm working on pnpm support for syncpack, @zkochan one thing I want to be clear on: does pnpm search each entry in pnpm-workspaces.yaml recursively, even if it doesn't end with `**`?

**Comment by zkochan** [comment]
maybe you can use this: https://github.com/Thinkmill/manypkg/tree/master/packages/get-packages

> does pnpm search each entry in pnpm-workspaces.yaml recursively, even if it doesn't end with **?

no

**Comment by aparajita** [comment]
> maybe you can use this: https://github.com/Thinkmill/manypkg/tree/master/packages/get-packages

That may be just what I'm looking for, thanks.

**Comment by aparajita** [comment]
@zkochan My modified version of syncpack is ready if you want to try it.

```sh
git clone https://github.com/aparajita/syncpack.git
cd syncpack
pnpm install
pnpm build
cd /path/to/pnpm-source
/path/to/syncpack/dist/bin.js list-mismatches
```

You may find the results interesting.

**Comment by zkochan** [comment]
I see. There are a few issues.

**Comment by aparajita** [comment]
Good incentive to implement package syncing in pnpm. 😁

**Comment by aparajita** [comment]
My PR to add pnpm support to [syncpack](https://github.com/JamieMason/syncpack) has been merged and is now available via `pnpm i -g syncpack`.

**Comment by evantahler** [comment]
Just chiming in that the linter idea would be great for @grouparoo!  On incoming PRs we could fail a test if the version of a dependency doens't match the rest of the monorepo.

**Comment by zkochan** [comment]
It is important to keep the versions of dependencies in sync, in a workspace. syncpack somewhat works but some builtin mechanics would be preferred. Maybe we could store all the version ranges in a single place (in the root of the workspace) and then just use something like `"dependencies": { "mongodb": "inherit" }` in the `package.json` files.

But even with linting configured, we currently may end-up with different versions installed in different packages, because when ranges are used, pnpm may install one version from the range in one project and some other version in another project. This issue will probably be solved in scope of https://github.com/pnpm/pnpm/issues/2222

**Comment by zkochan** [comment]
Although syncpack is pretty nice, it is hard to set it up. It is not easy to normalize all the version specs in workspace that has hundreds of packages. I think a more preferred solution would be to start with the critical dependencies, like React. So it would be nice to be able to force specific version ranges for specific dependencies.

I have added some new features to the experimental [meta-updater](https://github.com/pnpm/meta-updater) project that I have developed for keeping the pnpm monorepo in sync. Initially it could update the manifests but now it can also lint the `package.json` files without changing them.

With meta-updater we can program some smart algorithm for linting versions. For instance, we could have some json in the root of the repository that would declare the allowed version ranges for some dependencies, then version updater could use that information to for the listed ranges in each workspace project.

**Comment by wclr** [comment]
Some time ago decided that the appropriate method for the monorepo project that contains code for different services and apps (which are not packages supposed to be published) is to keep all the dependencies in one root package.json file, and actually only a single one.

There there is definitely a need for some tooling that will support this kind of workflow. For example to manage to deploys (with appropriate dependencies) or if you need to know which services use which deps, ect.. But the flat and plain model for all dependencies has significant advantages, everything becomes more transparent and pure.

Also as in my latest projects, I tend to use mostly PureScript, this model for managing npm dependencies is actually the only option that aligns well with all the dev setup (though for JS/TS bases projects I would too consider this model as a foundation).

Probably the mentioned approach with version `inheritance` can be useful in cases where nested package.json manifests are assumed.

**Comment by zkochan** [comment]
It makes sense to keep most of the dev dependencies in the root package.json only. And I think we should add some linting for this as well. If a dev dependency is in the root, it should not be in dev dependencies of non-root package.json files.

In other cases, if there are packages with prod dependencies and those package are published to the registry, then we should have their dependencies specified in their own package.json files.

**Comment by AlonMiz** [comment]
I've made a package for yarn for the exact same purpose. running it in our CI pipeline.
https://www.npmjs.com/package/yarn-one-version-rule
would be nice to have something similar for pnpm

**Comment by darkbasic** [comment]
I have a similar issue, but it might require a completely different solution.
I have a monorepo with several packages (`/packages/*`) and several examples to test these packages (`/examples/*`).
These packages need to share the exact same instance of some dependencies which will be provided by the user application (the examples in this case), thus I'm using peerDeps along with devDependencies. In the examples I have normal dependencies instead.
I make sure to specificy the same exact version in all the packages and all the examples, but this is not enough to ensure the same instance gets shared. In fact if the dependency itself has peer deps (like https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/package.json#L66) then pnpm will point the symlink to a different instance depending on which of its peer deps are present in the package.json of the package. This is not what would happen outside of the monorepo, because the example will provide the peerDep and the different packages will use it because they do not directly depend on it (they're just devDependencies).

I'll make an example:

**package A**:
```
  "peerDependencies": {
    "@mikro-orm/core": "^5.0.0-rc.1",
    [...]
  },
  "devDependencies": {
    "@mikro-orm/core": "5.0.0-rc.1",
    [...]
```

```
$ ls -al node_modules/\@mikro-orm/core
node_modules/@mikro-orm/core -> ../../../../node_modules/.pnpm/@mikro-orm+core@5.0.0-rc.1/node_modules/@mikro-orm/core
```

**package B**:
```
  "peerDependencies": {
    "@mikro-orm/cli": "^5.0.0-rc.1",
    "@mikro-orm/core": "^5.0.0-rc.1",
    "@mikro-orm/entity-generator": "^5.0.0-rc.1",
    "@mikro-orm/knex": "^5.0.0-rc.1",
    "@mikro-orm/migrations": "^5.0.0-rc.1",
    "@mikro-orm/postgresql": "^5.0.0-rc.1",
    "@mikro-orm/reflection": "^5.0.0-rc.1",
    [...]
  },
  "devDependencies": {
    "@mikro-orm/cli": "5.0.0-rc.1",
    "@mikro-orm/core": "5.0.0-rc.1",
    "@mikro-orm/entity-generator": "5.0.0-rc.1",
    "@mikro-orm/knex": "5.0.0-rc.1",
    "@mikro-orm/migrations": "5.0.0-rc.1",
    "@mikro-orm/postgresql": "5.0.0-rc.1",
    "@mikro-orm/reflection": "5.0.0-rc.1",
    [...]
```

```
$ ls -al node_modules/\@mikro-orm/core
node_modules/@mikro-orm/core -> ../../../../node_modules/.pnpm/@mikro-orm+core@5.0.0-rc.1_65b393e4e4022ad29afcf84ebe2e3777/node_modules/@mikro-orm/core
```

**example**:
```
  "dependencies": {
    "@mikro-orm/cli": "5.0.0-rc.1",
    "@mikro-orm/core": "5.0.0-rc.1",
    "@mikro-orm/entity-generator": "5.0.0-rc.1",
    "@mikro-orm/knex": "5.0.0-rc.1",
    "@mikro-orm/migrations": "5.0.0-rc.1",
    "@mikro-orm/postgresql": "5.0.0-rc.1",
    "@mikro-orm/reflection": "5.0.0-rc.1",
    [...]
```

```
$ ls -al node_modules/\@mikro-orm/core
node_modules/@mikro-orm/core -> ../../../../node_modules/.pnpm/@mikro-orm+core@5.0.0-rc.1_65b393e4e4022ad29afcf84ebe2e3777/node_modules/@mikro-orm/core
```

As you can see there are multiple instances of the same dependency:
```
$ ls -al node_modules/.pnpm | grep "@mikro-orm+core"
drwxr-xr-x 1 niko niko     24 Feb  1 12:17 @mikro-orm+core@5.0.0-rc.1
drwxr-xr-x 1 niko niko     24 Feb  1 12:17 @mikro-orm+core@5.0.0-rc.1_65b393e4e4022ad29afcf84ebe2e3777
```

Package B and the example share the same instance of `@mikro-orm/core` (because they have the same peer dependencies of `@mikro-orm/core` in their dependencies/devDependencies), while package A gets another instance because it doesn't.
How can I force all the packages to use the very same instance without having to add `@mikro-orm/entity-generator`, `@mikro-orm/knex`, `@mikro-orm/migrations`, `@mikro-orm/postgresql` and `@mikro-orm/reflection` to package A devDependencies? I tried overrides but they won't help (the version is already the same).
The library won't work if the instance isn't the same and thus running the examples from the monorepo will fail.

**Comment by JamieMason** [comment]
Take a look but syncpack versionGroups might be able to help you with that @darkbasic https://github.com/JamieMason/syncpack/releases/tag/6.1.0

**Comment by remorses** [comment]
I made a `.pnpmfile.cjs` to ensure some packages have only one version after every install

```js

const enforceSingleVersion = [
    'react-hot-toast', //
    'react',
    'react-dom'
]

function afterAllResolved(lockfile, context) {
    console.log(`Checking duplicate packages`)
    const packagesKeys = Object.keys(lockfile.packages);
    const found = {}
    for (let p of packagesKeys) {
        for (let x of enforceSingleVersion) {
            if (p.startsWith(`/${x}/`)) {
                if (found[x]) {
                    found[x] += 1
                } else {
                    found[x] = 1
                }
            }
        }
    }
    let msg = ''
    for (let p in found) {
        const count = found[p]
        if (count > 1) {
            msg += `${p} found ${count} times\n`
        }
    }
    if (msg) {
        throw new Error(msg)
    }
    return lockfile
}

module.exports = {
    hooks: {
        afterAllResolved
    }
}
```

**Comment by kevinkhill** [comment]
If you use `context.log` instead of `console.log` you get a nice label of where the log cam from :)

![image](https://user-images.githubusercontent.com/266076/183504347-32d2bd74-ab73-488d-9cc0-046af32131fe.png)

**Comment by evelant** [comment]
It would be conveneint to be able to specify package resolutions in package.json, somewhat like we can specify `"my-workspace-pkg": "workspace:*"`.

I imagine a really nice way to do this would be to use the same syntax. For example, in the workspace root package.json you could put `"@effect/core": "^0.0.13"` (I could imagine these in either a pnpm specific section or just dependencies) then in your packages write `"@effect/core": "workspace"` to mean "use whatever version is specific in workspace root". 

I think such a system would be convenient and consistent. `workspace` is already used to force resolution within a workspace. I think it would make sense to add the simple additional sematic "if a version is set to `workspace` it must have a version specified in the workspace root package.json, otherwise emit an error".

**Comment by zkochan** [comment]
@evelant it is probably a good idea but it should start with a protocol. "workspace" would mean a version published with the "workspace" dist-tag.

I think it might be fine to use "*" for this case as there is no legitimate reason to publish a package with "*" in the dependencies.

**Comment by mgambati** [comment]
@zkochan I liked the idea of the `inherit` or 'inherit: workspace' resolver more than the `*`, making it more explicit and not conflicting with the existing syntax (even if it should not be used).

**Comment by patroza** [comment]
@remorses https://github.com/pnpm/pnpm/issues/2713#issuecomment-1141000426
this is cool, thanks!

Im contemplating if adding a marker to package.json that we can pick up, would be a good solution.
`"shouldBeUnique": true` or so.

**Comment by aphex** [comment]
@zkochan did anything like `inherit` ever land for this? I think it is a solid idea when you want to sync packages up when the root workspace also requires the package. So root workspace `package.json` has `xyz:^1.0.4`, so any subpackage with `xyz: 'inherit'` would pick that up.  this would solve some current frustrations for me.

However this isn't a great setup for keeping package versions in sync that are not required in the root workspace `package.json`, its a bit harder to figure out which one to reference in this situation, but you could imagine a monorepo with a front-end and back-end workspace. Maybe you want to share react versions just in the front-end packages.

**Comment by dsilvasc** [comment]
Gradle does this with a "version catalog" file: https://docs.gradle.org/current/userguide/platforms.html#sub::toml-dependencies-format

A similar approach for pnpm might be:

versions.yaml or versions.toml:
```yaml
versions:
  react: ^18.0.2
packages:
  # react and react-dom versions should align
  react: react
  react-dom: react
  typescript: ^8.4.0
```

apps/myapp/package.json:
```json
{
  "dependencies": {
     "react": "catalog:*"
   },
  "devDependencies": {
      "typescript": "catalog:*"
  }
}
```

Where the `catalog` protocol looks at a versions file at the root of the repo, and `*` means use the default versions file (maybe you can have multiple).

**Comment by gluxon** [comment]
Hello fellow pnpm users! 👋

I wanted to share https://github.com/pnpm/rfcs/pull/1, which has a lot of the same ideas here. The RFC proposes a new `workspace-consistent:` protocol, which would behave similar to the prior mentioned `"inherits"` and `"catalog:*"` ideas. I only saw this issue today, but I find it cool how much of the ideas overlap.

The RFC adds a few features on top of a consistent root versions catalog.

- A change to the  `pnpm-lock.yaml` using these consistent version place holders can reduce `pnpm-lock.yaml` and `package.json` git merge conficts as these common versions are upgraded.
- Gives a mechanism for pnpm to know about and better enforce 1 version of a package transitively. (And not just direct project dependencies.)

The RFC got stalled a bit because of the complexity of the extra mile efforts above, but is something I'm still very invested in. Please feel free to drop ideas on the RFC or give suggestions.

(Disclaimer: I'm not a pnpm maintainer. Just a contributor that's submitted pull requests in the past.)

**Comment by zkochan** [comment]
In [Bit](https://bit.dev/) there is a concept known as environments (cc @GiladShoham). I think we can support something similar in pnpm. It would be not as powerful as in Bit but it could allow to sync dependencies between projects (even if they don't stay in the same project). For Bit it would also be beneficial as we would mention that the idea was borrowed from Bit and for more powerful envs people can use bit cli.

This is how I think it could work. In `package.json` we could add a new field. For instance:

```json
{
  "name": "foo",
  "version": "1.0.0",
  "pnpm": {
     "environment": "@comp/react-env@1.0.0"
  }
}
```

This field would contain the exact name and version of a package called an "environment package". This package would be a regular package with dependencies and peer dependencies in its package.json. Then we could add some fields to inherit fields from this env package. For instance:

```json
{
  "name": "foo",
  "version": "1.0.0",
  "pnpm": {
     "environment": "@comp/react-env@1.0.0",
     "dependenciesFromEnv": ["react"],
     "devDependenciesFromEnv": ["eslint"],
     "peerDependenciesFromEnv": ["react-dom"]
  }
}
```

pnpm would then lookup the versions of these dependencies in the env package.json and place them to regular dependencies fields. This is the package.json file of the `@comp/react-env@1.0.0`:

```json
{
  "name": "@comp/react-env",
  "version": "1.0.0",
  "dependencies": {
    "react": "16",
    "eslint": "4",
    "react-dom": "17",
    "jest": "22"
  },
  "peerDependencies": {
    "react-dom": "17"
    "react": "16"
  }
}
```

In memory, this is out project's package.json that will be updated with selected dependencies from the env:

```json
{
  "name": "foo",
  "version": "1.0.0",
  "dependencies": {
    "react": "16"
  },
  "devDependencies": {
    "eslint": "4"
  },
  "peerDependencies": {
    "react-dom": "17"
  },
  "pnpm": {
     "environment": "@comp/react-env@1.0.0",
     "dependenciesFromEnv": ["react"],
     "devDependenciesFromEnv": ["eslint"],
     "peerDependenciesFromEnv": ["react-dom"]
  }
}
```

**Comment by aphex** [comment]
@zkochan Heck ya now this is starting to look good. Could we use the the reference package name as a key though, so there could me multiple? Something like this.

```json
{
  "name": "foo",
  "version": "1.0.0",
  "pnpm": {
    "environment": {
      "@comp/react-env@1.0.0": {
        "dependencies": ["react"],
        "devDependencies": ["eslint"],
        "peerDependencies": ["react-dom"]
      }
    }
  }
}
```

If someone adds the same dependency from two different reference packages I would say we just use a "last in wins" strategy.

**Comment by zkochan** [comment]
I don't know if it makes sense to support multiple. This is inspired by Bit, which supports one env per component. I think one env should be enough because you can create envs that inherit other envs.

**Comment by zkochan** [comment]
> If I am following you would solve this by creating new packages in the mono repo that are just used as a environment for other packages? Just a collection of all the packages within the monorepo that should have consistent versioning?

yes, I think so

**Comment by aphex** [comment]
Alright I am following ya now. That's not bad. I think my issue is personal aesthetics, which clearly is not a solid reason to change anything haha. 

Something about 4 sibling properties in an object all with "env" attached to them to indicate they are related vs keeping them all tucked away inside a "env" property. 🤷‍♂️ 

Either way getting a feature like this would help a lot, whichever way you prefer to configure it :)

**Comment by evelant** [comment]
@aphex it feels a little odd at first because in most setups there's a lot of friction to adding new packages in a monorepo but having separate packages for envrionments required by your other packages is exactly what Bit does and after spending some time with it I think I can say it's a very useful pattern. Bringing some of that funcionality into pnpm IMO would be great.

**Comment by aphex** [comment]
@evelant No I get it, I am onboard and definitely want to see the functionality. If Bit has a solid pattern it should be followed. My aesthetic thoughts on the config design are not critical here.

**Comment by zkochan** [comment]
> Something about 4 sibling properties in an object all with "env" attached to them to indicate they are related vs keeping them all tucked away inside a "env" property.

I don't know how this will be configured and how the properties will be named. It is just an example to explain the concept. Bit doesn't have such properties at all because it can automatically detect what packages from the env are used by the component.

**Comment by sneko** [comment]
Until there is an automatic way to find out duplicated packages I listed some steps that helped me: https://github.com/pnpm/pnpm/discussions/6055#discussioncomment-5018827

**Comment by dtothefp** [comment]
I've recently started to use PNPM in a monorepo and it solves a lot of problems but at the same time I'm experiencing this issue with singleton packages. I modified the `afterAllResolved` solution above and found that I have 392 duplicated dependencies and this is in a monorepo that is only moderate in size. For NodeJS projects maybe this is not a problem but for tools such as Eslint and for module bundlers like Webpack this presents issues. Is the only solution to manually find these duplicated dependencies and then specify the `pnpm.overrides` in `package.json`?

**Comment by neongreen** [comment]
> I made a `.pnpmfile.cjs` to ensure some packages have only one version after every install [...]

@remorses's solution is great, but unfortunately only allows checking for the entire workspace. Sometimes you care about individual projects in the workspace having singleton packages (eg. `@emotion`), but it's ok for two different projects in the workspace to have different emotion versions.

**Comment by zkochan** [comment]
This feature is close to completion thanks to @gluxon. You can already try it using the branch from the next PR: https://github.com/pnpm/pnpm/pull/8122

Just place your versions in the catalog section of `pnpm-workspace.yaml`. For instance:

```yaml title="xxx"
catalog:
  react: 16
```

Then in your `package.json` use the `catalog:` version:

```json
{
  "dependencies": {
    "react": "catalog:"
  }
}
```

If you check it out please leave your experiences. We want to release it soon.

**Comment by remorses** [comment]
Catalogs don't really solve this problem, usually i get multiple instances of `next` for example when they depend on peer dependencies that resolve to different versions based on which workspace it is installed (or if it is a transitive dependency on the parent dependency).

For example when i update `next` i will also have to update all of it's peer dependencies and make sure pnpm doesn't duplicate the package in pnpm-lock.yaml

Here is an example of next getting duplicated because babel-plugin-macros resolves to different versions:

```yml
    next@14.2.4(@babel/core@7.24.6)(@playwright/test@1.43.1)(babel-plugin-macros@2.8.0)(react-dom@18.3.1(react@18.3.1))(react@18.3.1):
    dependencies:
      '@next/env': 14.2.4
      '@swc/helpers': 0.5.5
      busboy: 1.6.0
      caniuse-lite: 1.0.30001638
      graceful-fs: 4.2.11
      postcss: 8.4.31
      react: 18.3.1
      react-dom: 18.3.1(react@18.3.1)
      styled-jsx: 5.1.1(@babel/core@7.24.6)(babel-plugin-macros@2.8.0)(react@18.3.1)
    optionalDependencies:
      '@next/swc-darwin-arm64': 14.2.4
      '@next/swc-darwin-x64': 14.2.4
      '@next/swc-linux-arm64-gnu': 14.2.4
      '@next/swc-linux-arm64-musl': 14.2.4
      '@next/swc-linux-x64-gnu': 14.2.4
      '@next/swc-linux-x64-musl': 14.2.4
      '@next/swc-win32-arm64-msvc': 14.2.4
      '@next/swc-win32-ia32-msvc': 14.2.4
      '@next/swc-win32-x64-msvc': 14.2.4
      '@playwright/test': 1.43.1
    transitivePeerDependencies:
      - '@babel/core'
      - babel-plugin-macros

  next@14.2.4(@playwright/test@1.43.1)(babel-plugin-macros@3.1.0)(react-dom@18.3.1(react@18.3.1))(react@18.3.1):
    dependencies:
      '@next/env': 14.2.4
      '@swc/helpers': 0.5.5
      busboy: 1.6.0
      caniuse-lite: 1.0.30001638
      graceful-fs: 4.2.11
      postcss: 8.4.31
      react: 18.3.1
      react-dom: 18.3.1(react@18.3.1)
      styled-jsx: 5.1.1(babel-plugin-macros@3.1.0)(react@18.3.1)
    optionalDependencies:
      '@next/swc-darwin-arm64': 14.2.4
      '@next/swc-darwin-x64': 14.2.4
      '@next/swc-linux-arm64-gnu': 14.2.4
      '@next/swc-linux-arm64-musl': 14.2.4
      '@next/swc-linux-x64-gnu': 14.2.4
      '@next/swc-linux-x64-musl': 14.2.4
      '@next/swc-win32-arm64-msvc': 14.2.4
      '@next/swc-win32-ia32-msvc': 14.2.4
      '@next/swc-win32-x64-msvc': 14.2.4
      '@playwright/test': 1.43.1
    transitivePeerDependencies:
      - '@babel/core'
      - babel-plugin-macros
```

It would be great if there was a way to tell pnpm to install next only once and ignore different peer dependencies resolutions

**Comment by zkochan** [comment]
:ship: [9.5.0](https://github.com/pnpm/pnpm/releases/tag/v9.5.0-beta.0)

@remorses your issue is unrelated to this issue.
